### ISSUE TITLE
Check configuration inputs for Run 1.1

### ISSUE THREAD

**Issue by cwwalter** [LSSTDESC/DC2_Repo#66]
In PR #63 we worked on a PhoSim instance catalog that is now being tested in Run 1.0.

Before we start R1.1 there are various configuration options that we should review and check to make sure the current values are consistent with our most recent understanding.  

Examples include: tree rings parameters in phosim which are in the silicon.txt file, x-talk and hot pixel parameters etc., etc

Perhaps first we need to summarize what we need to look at.

**Comment by johnrpeterson** [comment]
i'm not sure what this is specifically about.  this sort of thing is periodically reviewed in the project, and gets updated in every release.  if you find something specific, certainly issue a phosim ticket.

**Comment by cwwalter** [comment]
For example, we have new recent measurement from the SAWG for the tree ring parameters.  Should we (or have you already) updated the PhoSim configs?

I want to check off on everything before we burn > 10M hours...

Having a summarized list would be good.  It sounds like you might have a list of recently updated parameters and what you changed?  Could you share it?  That sounds like it might really help.

**Comment by johnrpeterson** [comment]
there are hundreds of changes with each release, Chris.....

but for tree rings, do you have the relative intensity compared to what was measured by Andrei before (~1 year ago)?

**Comment by nomerot** [comment]
Hyeyun updated on the ILT tree rings at the last SAWG meeting https://confluence.slac.stanford.edu/display/LSSTDESC/July+-+December+2017, she also provided txt files with TR amplitude for a few ILT sensors as examples. e2v TRs are negligible. John, could you remind how you define the TR profile, was it a superposition of a few sines?

**Comment by katrinheitmann** [comment]
| Effect 	| PhoSim setting 	|  DM Status  | Customer | Ready for Run 1.1 |
|:-:	|:-:	|:- |:- |
| Tree rings 	|   	? |   ?         | ?  |  ?            |

**Comment by cwwalter** [comment]
> there are hundreds of changes with each release, Chris.....

Dear John,

Sorry I was traveling today and am now in California for the holiday.

OK; so let me be a bit more specific.  Of course there are many changes in each release. Hopefully you are tracking those yourself in a detailed release notes.

I am talking about something more specific.  We want to document the parameters in the instrument directory we are using for the run.   We need to be able to document and explain the parameters we have used for things like hot pixels, tree rings, x-talk etc. These are the parameters in the segmentation.txt file, the silicon.txt file etc.  We need to document these so that we can:

- Report on what we simulated accurately
- Make sure we are using the most appropriate values
- Have this information available to any people who need it when they are using the files for analysis.

Of course sometime changes to the underlying model might also be appropriate to report if they are not accurately represented in the previous published PhoSim information. Also,  it might be useful to explain items such as the model the tree ring parameters are used in.

I am now asking for the help of the PhoSim team in generating this list.  Katrin started with a suggestion of a table above. Happy holidays to all!

**Comment by johnrpeterson** [comment]
Chris, we produced something like that for the project review of phosim in 2013.  So this culminated in the 170 page phosim reference document, but it is out of date in almost every area.  its really a daunting task to update it.   i remember it took 6 or 9 months of work, and we did this a couple years before this.   since then we have largely added detail in an iterative manner in the project and resolved details informally.  i would like to update this for many reasons (not just DESC), but what timescale are you even thinking of?  It some cases, it is simple (e.g. numbers of pixels), but most cases it is very complicated to document this properly.

Andrei, it looks like if a simulate a flat crudely, i get about 0.6 or 0.7% variation, so its probably already set to roughly the right level already.  if you provide more details from HyeYun Park's plots, we can simulate the same thing and be more precise.

**Comment by cwwalter** [comment]
> Chris, we produced something like that for the project review of phosim in 2013. So this culminated in the 170 page phosim reference document, but it is out of date in almost every area. its really a daunting task to update it. i remember it took 6 or 9 months of work, and we did this a couple years before this. since then we have largely added detail in an iterative manner in the project and resolved details informally. i would like to update this for many reasons (not just DESC), but what timescale are you even thinking of? It some cases, it is simple (e.g. numbers of pixels), but most cases it is very complicated to document this properly.

Hi John,

As I explained above I am not asking for a complete validation / documentation of the PhoSim models.  I understand this is a a big job.  What we need is much more restricted. For the parameters that the user can in principle change in the instrument directory we want to write down what parameters were used for the simulation.  The ones that come to mind for me includes:

- Hot pixel rate
- X-Talk matrix
- Tree ring intensity / pattern
- Sensor choice (I mean E2V vs ITL and associated geometry).
- Versions of the system throughputs 

These are things we might in principle change.  They are all in the instrument directory.  We don't need everything in the directory.  We wouldn't touch the optics/perturbation/actuator/filter models for example.

If you could just start and let us know if there are others you think we should consider that would be good. I think we might want to also write down the items related to dark current, gain etc (I have had people working with DC1 ask me about/need these values for their analysis who thought we were using the amp files) but it would be helpful for the SAWG people to comment. We should work on this together.  I don't think this is a big job.  It is mostly documenting things that are in a few files so we can review and document them for the people who will be using DC2.

**Comment by johnrpeterson** [comment]
well, if you want to just worry about those 5 things then that is ok. 

these three make sense:  hot pixel rate & cross-talk is based off of prototype devices, so i wouldn't worry about those.  the default is to use ITL (e2v is implemented as well if you wanted to use that instead for some reason -i lsst_e2v ).

these two don't make sense:   phosim doesn't use any version of system throughputs.  and tree rings are based on physics.

**Comment by nomerot** [comment]
for the tree rings we would need to define the amplitude, period and geometry (location of the center, I believe it should be fixed now but there are 4 possible orientations). The amplitude and period should be defined for each sine in case we have more than one (and a phase shift between them?).

**Comment by ivezic** [comment]
@johnrpeterson I don't understand this statement "phosim doesn't use any version of system throughputs". E.g., how do you treat mirror reflectivity, lenses, filter bandpasses, etc?

**Comment by johnrpeterson** [comment]
andrei, the apparent tree rings variation in flats, for instance, are not simple functions since you have to forward simulate them and then see what the amplitude, pattern, etc. is.  so you could make a model based on what you see in the flats, for instance, but its not trivial to predict.  we need to get all the details right about the tree ring physics.

zeljko, it is based on various photon physics.  so angle-dependent wavelength-dependent filter functions which then couples with the raytracing or the conversion into photo-electron which depends on all sorts of details, etc.  the system throughputs is something we use in the validation framework code, but is not a phosim input.

this illustrates what is getting confused on this thread.  y'all are looking more at predicted observable outputs and thinking those are the same as the inputs.  the two are the same for very simple things like numbers of pixels for a chip or if you make "toy" simulations, but not for correct simulations. consequently, it is more well-defined to define what you observe in real life and then compare output measurements of that.

**Comment by ivezic** [comment]
@johnrpeterson Thanks for the answer - but I am still puzzled (my question was not
explicit enough). What I want to know is how do we ensure that what's in phosim is
consistent with the properties of LSST system. For example, in case of "angle-dependent wavelength-dependent filter functions" you mentioned, we need to ensure that your inputs 
for those functions correspond to our devices as delivered (or as designed at this time) 
if we want to reliably model ghosts, etc. How do we do version control of all the phosim 
inputs that define/describe the LSST system phosim is trying to model? Are the inputs
that phosim takes delivered by the SysEng team, Engineering & Facility db, or something
else? E.g., if the filter functions change, how does that change propagate to phosim?

**Comment by nomerot** [comment]
we studied how to relate the variation of doping and TR amplitude in flats in Phosim, this is documented in B. Beamer et al 2015 JINST 10 C05027 so this should be straightforward. the period and amplitude in the production ITL sensors can be taken from H.Y. Park et al 2017 JINST 12 C05015. we still need to decide on the TR orientation in the rafts and how many sines are used to avoid TR to be too periodical.

**Comment by cwwalter** [comment]
For the tree rings: As Andrei notes above the doping parameters and pattern are set in the silicon.txt file including the period etc.  We understand that this does not exactly tell you the resulting output but we want to know what pattern was put in and what the resulting output variation looks like after the run (See also the reference to the study by Ben B. and Andrei). This is important for us to understand when we are evaluating the correction algorithms, comparing with test bench data etc.

For the throughputs:  we understand that there are complicated interactions between the pieces but we want to document (for example) what version of the filter transmission curves are used as the input.  Zeljko expanded on this above.

We can really use the help of you and the PhoSim team on all of this.

**Comment by johnrpeterson** [comment]
zeljko-  in system engineering we compare to throughput curves and define an accuracy tolerance.  that is one of the hundreds of validation tests that gets versioned and automatically run constantly.  if any change to the physics or input data violates the tolerance then we'd know about it instantly.  before any formal version release, then we'd have to resolve it.

chris/andrei-  the pattern is certainly different depending on wavelength and  whether you a looking at flats or astrometry or psf patterns, etc.  and yes, the overall level was set with Ben already, remember?  there have been some improvements since then as well.  but what i am saying is the best way to determine what it looks like is to actually simulate something (e.g. flats).  that's how i got the 0.6% or 0.7% i referenced earlier.

**Comment by ivezic** [comment]
@johnrpeterson When you say "system engineering", you mean the LSST System Engineering, right?
I am in the middle of a process of updating system properties with Steve Ritz and Chuck Claver (in the context of updates to the overview paper). We have github repos that contain "official" versions of our knowledge about the LSST system (together with the EFD). What I would like to confirm is that you compare to the same repos, etc. - how can I find out what exactly you do (what tools, where do you get LSST info, etc) when you validate phosim?

**Comment by johnrpeterson** [comment]
zeljko, it could be out of date but not sure.  lets go through the project processes stuff offline.

Chris/Andrei-  Andrei & Ben tuned some of the value already and we added all sorts of details to make it realistic in future versions, so now i'm not sure what you are asking for?  if you don't need to know exactly what it looks like, and just approx. then we all already know that.

**Comment by cwwalter** [comment]
Hi All,

I want to come back to this before we start R1.1.  Obviously, we need these values for documentation and use. But we also need to understand and possibly set some values before we can even start. I've mentioned x-talk etc above but I particularly wanted to expand on the hot pixel rate and write this down so it isn't forgotten.

For the amplifier files from either simulation we need  and want to run ISR (this is one of the things we are learning about) and this means figuring out how to do hot pixel masking etc.  The way PhoSim does this is you set a rate and then it chooses random pixels (deterministically) to make hot pixels.  But, @jchiang87 has suggested the best way for us to do handle this might be to actually make masks ourselves and then just apply those (without turning the hot pixels on).  He recommended generating our own masks from scratch, including the edge rolloff regions, midline mask for the bloom stop, and maybe random columns and pixels to account for defects.  

So, the way we would do this is by turning off the hot pixel rate in the segmentation.txt file and just applying the masks. DM will assume the camera team will give them masks so this also saves us the step of writing code to try to extract them from the amp files.

**Comment by johnrpeterson** [comment]
Hot pixels though are really easy though.  There aren't that many per chip (maybe just ~2).   And you can find them trivially in the bias frames.  I think it would just a tiny piece of code to write to find them and provide it to DM.

**Comment by SimonKrughoff** [comment]
Cross posted from #111:

> use:
> pixelerror 0
> if you don't want the PRNU.

**Comment by katrinheitmann** [comment]
In view that it is not a good idea to make changes in thoroughly tested input files just before we start Run 1.1 and keeping in mind that the main purpose for Run 1.1 is to test all of our configurations I propose to close this issue now (it's title is Run 1.1). The hot pixels are turned on. @cwwalter  -- if you prefer them turned off and we do not have to do anything besides setting the appropriate variable then John can quickly send the correct setting for that. All the input files for PhoSim are in our run directory at NERSC and listed and documented in the google doc. The effects we turned off are listed there as well. @jchiang87 @TomGlanzman we should be good to go with regard to just this part, correct (the phosim input settings)? Thanks everybody!

**Comment by cwwalter** [comment]
Given the situation, I agree we should not hold up starting 1.1 for this. 

I think the only real configuration issue is hot pixels and columns as you say. Leaving them on means that an extra step is needed  into the workflow chain to create masks to use the amplifier files, and we can't (at least just) use the technique Jim suggested of creating our own masks with the features we think are relevant from our real sensors.  So, I'm interested in Jim's opinion. Turning it off would require changing column values in the segmentation.txt file in the lsst instrument directory. 

But, please don't just close this issue.  I think we might want to either rename it or copy the information over to a new issue but the issues and discussion here are both important and unresolved.

**Comment by jchiang87** [comment]
I'm thinking that the step of generating masks to handle the hot pixels that phosim adds will be a small perturbation to the work of generating the other masks we'll need, i.e., edge rolloff, midline stop, so I'm not very concerned about turning off the hot pixels in the phosim configuration.  TBH, my earlier comment about just accounting for those sorts of defects by generating the masks without adding the hot pixels really pertained to the imsim runs.  I didn't want to go to the bother of adding those sorts of defects to the imsim code if we were just going to mask them anyways.

**Comment by cwwalter** [comment]
> I'm thinking that the step of generating masks to handle the hot pixels that phosim adds will be a small perturbation to the work of generating the other masks we'll need, i.e., edge rolloff, midline stop, so I'm not very concerned about turning off the hot pixels in the phosim configuration. 

OK thanks. Yes, then I agree we are ready.
